s5_base/run.sh

```diff
@@ -6,148 +6,271 @@
+    local/samromur_data_prep.sh $samromur_root dev data
+    local/samromur_data_prep.sh $samromur_root eval data
+    
+    utils/fix_data_dir.sh data/train
+    utils/fix_data_dir.sh data/dev
+    utils/fix_data_dir.sh data/eval
 fi
 
 # Prepare features
 if [ $stage -le 1 ]; then
-  echo "Make mfccs"
-  # Make MFCCs for each dataset
-  for name in train dev eval; do
-    steps/make_mfcc.sh --mfcc-config conf/mfcc.conf \
-      --nj ${num_jobs} --cmd "$train_cmd --max-jobs-run 99" \
-      data/${name} exp/make_mfcc/${name} $mfccdir
-    utils/fix_data_dir.sh data/${name}
-  done
+    echo "Make mfccs"
+    # Make MFCCs for each dataset
+    for name in train eval dev; do
+        steps/make_mfcc.sh \
+        --mfcc-config conf/mfcc.conf \
+        --nj ${num_jobs} \
+        --cmd "$train_cmd" \
+        data/${name} exp/make_mfcc $mfccdir \
+        || error 1 "Failed creating MFCC features";
+    done
 fi
 
 if [ $stage -le 2 ]; then
-  echo "Comute cmvn"
-  for name in train dev eval; do
-    steps/compute_cmvn_stats.sh data/${name} exp/make_mfcc/${name} $mfccdir
-    utils/fix_data_dir.sh data/${name}
-    utils/validate_data_dir.sh data/${name}
-  done
+    echo "Comute cmvn"
+    for name in train eval dev; do
+        steps/compute_cmvn_stats.sh \
+        data/${name} exp/make_mfcc $mfccdir;
+        
+        utils/validate_data_dir.sh data/"$name" \
+        || utils/fix_data_dir.sh data/"$name" || exit 1;
+    done
 fi
 
 if [ $stage -le 3 ]; then
-  echo "Prepare dict and lang directories"
-  local/prep_samromur_lang.sh data/train/text data
+    echo "Identify OOV words"
+    cut -d' ' -f2- data/train/text \
+    | tr ' ' '\n' | sort -u \
+    > data/train/wordlist.txt || exit 1;
+    
+    comm -23 data/train/wordlist.txt \
+    <(cut -f1 $prondict_orig | sort -u) \
+    > data/train/oov_wordlist.txt || exit 1;
+    
+    echo "Use a grapheme-to-phoneme model to generate the pronunciation of OOV words in the training set"
+    g2p.py --apply data/train/oov_wordlist.txt --model $g2p_model \
+    --encoding="UTF-8" > data/train/oov_with_pron.txt || exit 1;
+    
+    echo "Add the OOV words to the prondict"
+    cat $prondict_orig data/train/oov_with_pron.txt | sort -k1,1 | uniq > "$prondict" || exit 1;
+    
+    for n in eval dev; do
+        nb_tokens=$(cut -d' ' -f2- data/$n/text | wc -w)
+        cut -d' ' -f2- data/$n/text \
+        | tr ' ' '\n' | sort |uniq -c \
+        > data/$n/words.cnt || exit 1;
+        
+        comm -23 <(awk '$2 ~ /[[:print:]]/ { print $2 }' data/$n/words.cnt | sort) \
+        <(cut -f1 $prondict | sort -u) > data/$n/vocab_text_only.tmp
+        nb_oov=$(join -1 1 -2 1 data/$n/vocab_text_only.tmp <(awk '$2 ~ /[[:print:]]/ { print $2" "$1 }' data/$n/words.cnt | sort -k1,1) \
+        | sort | awk '{total = total + $2}END{print total}')
+        oov=$(echo "scale=3;$nb_oov/$nb_tokens*100" | bc)
+        echo "The out of vocabulary rate for $n is: $oov" >> oov_rate || exit 1;
+    done
```

Comment:
Can you add a comment to this section explaining what's going on? It's kinda explained at the end but it'd be good to have an explanation at the beginning too.